pkg/controller/resourceclaim/controller.go

```diff
@@ -0,0 +1,393 @@
+	sep := strings.Index(key, ":")
+	if sep < 0 {
+		return fmt.Errorf("unexpected key: %s", key)
+	}
+	prefix, object := key[0:sep+1], key[sep+1:]
+	namespace, name, err := cache.SplitMetaNamespaceKey(object)
+	if err != nil {
+		return err
+	}
+
+	switch prefix {
+	case podKeyPrefix:
+		return ec.syncPod(ctx, namespace, name)
+	case claimKeyPrefix:
+		return ec.syncClaim(ctx, namespace, name)
+	default:
+		return fmt.Errorf("unexpected key prefix: %s", prefix)
+	}
+
+}
+
+func (ec *resourceClaimController) syncPod(ctx context.Context, namespace, name string) error {
+	logger := klog.FromContext(ctx)
+	pod, err := ec.podLister.Pods(namespace).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.V(5).Info("nothing to do for pod, it is gone", "pod", klog.KRef(namespace, name))
+			return nil
+		}
+		return err
+	}
+
+	// Ignore pods which are already getting deleted.
+	if pod.DeletionTimestamp != nil {
+		logger.V(5).Info("nothing to do for pod, it is marked for deletion", "pod", klog.KRef(namespace, name))
+		return nil
+	}
+
+	for _, podClaim := range pod.Spec.ResourceClaims {
+		if err := ec.handleClaim(ctx, pod, podClaim); err != nil {
+			ec.recorder.Event(pod, v1.EventTypeWarning, "ResourceClaimCreation", fmt.Sprintf("PodResourceClaim %s: %v", podClaim.Name, err))
+			return fmt.Errorf("pod %s/%s, PodResourceClaim %s: %v", namespace, name, podClaim.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// handleResourceClaim is invoked for each volume of a pod.
+func (ec *resourceClaimController) handleClaim(ctx context.Context, pod *v1.Pod, podClaim v1.PodResourceClaim) error {
+	logger := klog.FromContext(ctx)
+	logger.V(5).Info("checking", "podClaim", podClaim.Name)
+	if podClaim.Source.Template == nil {
+		return nil
+	}
+
+	claimName := resourceclaim.Name(pod, &podClaim)
+	claim, err := ec.claimLister.ResourceClaims(pod.Namespace).Get(claimName)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if claim != nil {
+		if err := resourceclaim.IsForPod(pod, claim); err != nil {
+			return err
+		}
+		// Already created, nothing more to do.
+		logger.V(5).Info("claim already created", "podClaim", podClaim.Name, "resourceClaim", claimName)
+		return nil
+	}
+
+	// Create the ResourceClaim with pod as owner.
+	isTrue := true
+	claim = &v1.ResourceClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: claimName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion:         "v1",
+					Kind:               "Pod",
+					Name:               pod.Name,
+					UID:                pod.UID,
+					Controller:         &isTrue,
+					BlockOwnerDeletion: &isTrue,
+				},
+			},
+			Annotations: podClaim.Source.Template.Annotations,
+			Labels:      podClaim.Source.Template.Labels,
+		},
+		Spec: podClaim.Source.Template.Spec,
+	}
+	metrics.ResourceClaimCreateAttempts.Inc()
+	_, err = ec.kubeClient.CoreV1().ResourceClaims(pod.Namespace).Create(ctx, claim, metav1.CreateOptions{})
+	if err != nil {
```

Comment:
If this occurs, then perhaps the controller should use a MutationCache. I think I prefer to keep the error for now and later decide whether it should be handled differently because keeping the pod in the queue and trying again is the right thing to do.